include/rdma/fi_prov.h

```diff
@@ -128,6 +131,17 @@ fi_param_get_bool(struct fi_provider *provider, const char *param_name, int *val
 	return fi_param_get(provider, param_name, value);
 }
 
+/* Prepend DIRECT_FN to provider specific API functions for global visibility
+ * when using fabric direct.  If the API function is static use the STATIC
+ * macro to bind symbols globally when compiling with fabric direct.
+ */
+#ifdef FI_DIRECT_ENABLED
+#define DIRECT_FN __attribute__((visibility ("default")))
+#define STATIC
+#else
+#define DIRECT_FN
+#define STATIC static
```

Comment:
The STATIC definition does not need to be here.  This definition can be placed elsewhere such as gnix.h if desired.  It was placed here in the hope that other providers could use it when implementing fi direct.  However, other providers may find a better solution and ignore this STATIC definition.  From what hppritcha, jswaro and I found, applications cannot compile with fi direct enabled when the provider's api functions are defined as static since the symbols for those functions don't have global bindings in libfabric.a then.

> On Feb 16, 2016, at 9:24 AM, Sung-Eun Choi <notifications@github.com> wrote:
> 
> In include/rdma/fi_prov.h:
> 
> > @@ -128,6 +131,17 @@ fi_param_get_bool(struct fi_provider *provider, const char *param_name, int *val
> >  	return fi_param_get(provider, param_name, value);
> >  }
> >  
> > +/* Prepend DIRECT_FN to provider specific API functions for global visibility
> > + * when using fabric direct.  If the API function is static use the STATIC
> > + * macro to bind symbols globally when compiling with fabric direct.
> > + */
> > +#ifdef FI_DIRECT_ENABLED
> > +#define DIRECT_FN __attribute__((visibility ("default")))
> > +#define STATIC
> > +#else
> > +#define DIRECT_FN
> > +#define STATIC static
> Does this definition of STATIC have to be here? Seems like it should be up to the provider whether the functions are static if FI_DIRECT is not being used.
> 
> —
> Reply to this email directly or view it on GitHub.
>